lib/source/mqtt/aws_iot_mqtt_operation.c

```diff
@@ -580,25 +595,65 @@ void AwsIotMqttInternal_DestroyOperation( void * pData )
     if( ( pOperation->operation == AWS_IOT_MQTT_PUBLISH_TO_SERVER ) &&
         ( pOperation->pPublishRetry != NULL ) )
     {
-        IotMutex_Lock( &( pOperation->pMqttConnection->timerMutex ) );
+        if( lockMutexes == true )
+        {
+            IotMutex_Lock( &( pMqttConnection->timerMutex ) );
+        }
 
-        /* Remove the timer event from the timer event list before freeing it.
-         * The return value of this function is not checked because it always
-         * equals the publish retry event or is NULL. */
-        ( void ) IotListDouble_RemoveFirstMatch( &( pOperation->pMqttConnection->timerEventList ),
-                                                 NULL,
-                                                 NULL,
-                                                 pOperation->pPublishRetry );
+        /* Remove the timer event from the timer event list before freeing it. */
+        if( IotLink_IsLinked( &( pOperation->pPublishRetry->link ) ) == true )
+        {
+            IotListDouble_Remove( &( pOperation->pPublishRetry->link ) );
+        }
 
-        IotMutex_Unlock( &( pOperation->pMqttConnection->timerMutex ) );
+        if( lockMutexes == true )
+        {
+            IotMutex_Unlock( &( pMqttConnection->timerMutex ) );
+        }
 
         AwsIotMqtt_FreeTimerEvent( pOperation->pPublishRetry );
     }
 
-    /* Check if a wait semaphore was created for this operation. */
+    /* Check if this is a waitable operation. */
     if( ( pOperation->flags & AWS_IOT_MQTT_FLAG_WAITABLE ) == AWS_IOT_MQTT_FLAG_WAITABLE )
     {
+        /* Destroy wait semaphore. */
         IotSemaphore_Destroy( &( pOperation->notify.waitSemaphore ) );
+
+        /* Decrement the MQTT connection's count of waitable operations. Ignore
+         * operations which are tied to the connection and do not modify
+         * the reference count: CONNECT, DISCONNECT, and PINGREQ. */
+        if( ( pOperation->operation != AWS_IOT_MQTT_CONNECT ) &&
+            ( pOperation->operation != AWS_IOT_MQTT_DISCONNECT ) &&
+            ( pOperation->operation != AWS_IOT_MQTT_PINGREQ ) )
+        {
+            if( lockMutexes == true )
```

Comment:
This code is repeated elsewhere as well, so an helper method such as _decreaseConnectionRefCount would be good.